### PR TITLE
#165305754 Fix validations errors on signup

### PR DESCRIPTION
**Pull request by mdn-network**
[Fixes 165305754]
- Fix validation errors on signup
#### Description of Task to be completed?
- Fix password validations
- Fix error messages username and email
- Fix failing tests
#### How should this be manually tested?
- `$ git clone https://github.com/andela/ah-the-immortals-backend.git`
- `$ git fetch origin` - To get all branches
- `$ bg-fix-signup-validation-errors-165305754`
- Create a virtual environment and activate it
        `$ python3.7 venv venv`
        ` $ source venv/bin/activate`
- Add a .env file as in the .env.example
        `$ source .env`
- ` $ pip intsall -r requirements.txt`
- Start server ` $ python manage.py runserver`
- Test the following endpoints on postman
   - Sign in `/api/users/`
- Run tests ` $ python api/manage.py test`
#### What are the relevant pivotal tracker stories?
[#165305754](https://www.pivotaltracker.com/n/projects/2326460/stories/165305754)
#### Screenshots
image

image

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -16,20 +16,63 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+        if stats.letters_uppercase < 2:
+            errors.append("password should have at least of 2 uppercase letters")
+
+        if stats.letters_lowercase < 3:
+            errors.append("password should have at least 3 lowercase letters")
+
+        if stats.numbers < 2:
+            errors.append("password should have at least 2 digits")
+
+        if stats.special_characters < 1:
```
Comment (Ogutu-Brian) [inline authors/apps/authentication/serializers.py:55]: Hello @Philipotieno ,
This part of your code is not tested. You need to add unit tests for it

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -16,20 +16,63 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+            errors.append("password should have at least 1 special character")
+
+        if stats.sequences_length > 2:
```
Comment (Ogutu-Brian) [inline authors/apps/authentication/serializers.py:66]: This part is also untested

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -16,20 +16,63 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+                "password": list(errors)
+            })
+
+    def get_password_field_erros(self, value):
+        """
+        Method to get null, blank and empty fields for password
+        """
+        if not value.get("password"):
+            raise serializers.ValidationError({
+                "password": "Provide your password"
```
Comment (Ogutu-Brian) [inline authors/apps/authentication/serializers.py]: This part is also untested. Would you test all the new code that this PR is adding to the codebase

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]

**Comment by Ogutu-Brian** [comment]
The method `get_password_policy_rrors` has a lot of if statements, the maximum cognitive depth should be 5, in accordance with code climate report. You could refactor it to reduce the code-smells thrown by code-climate

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py]: I have tested it now

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py:66]: I have written tests to cover that

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py:55]: Tested this one too

**Review by Ogutu-Brian — APPROVED** [review]
LGTM

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -16,20 +16,53 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
-            'max_length': 'Password cannot be more than 30 characters',
-            'invalid': 'Password must have alphanumeric characters with no space',
-        }
+        allow_null=True,
+        allow_blank=True,
+        required=True
     )
 
+    def validate(self, value):
+        self.get_password_policy_rrors(value.get("password"))
```
Comment (oluwagbenga-joloko) [inline authors/apps/authentication/serializers.py]: can you correct this`rrors` to `errors` here?

File: authors/apps/authentication/serializers.py [hunk 5]
```diff
@@ -16,20 +16,53 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+
+    def get_password_policy_rrors(self, password):
+        """
+        Method to captures password policy errors
+        """
+        stats = PasswordStats(password)
+        errors = []
+        if stats.letters_uppercase < 2:
+            errors.append(
+                "password should have at least of 2 uppercase letters")
```
Comment (oluwagbenga-joloko) [inline authors/apps/authentication/serializers.py]: This should be `password should have at least 2 uppercase letters`

File: authors/apps/authentication/serializers.py [hunk 6]
```diff
@@ -44,7 +77,7 @@ class RegistrationSerializer(serializers.ModelSerializer):
             )
         ],
         error_messages={
-            'invalid': 'Email must be of the format name@domain.com and should not contain any special characters before @',
```
Comment (oluwagbenga-joloko) [inline authors/apps/authentication/serializers.py:47]: I think you should use the default email field. The message should be 'the email address provided is invalid'. refer to the documentation below. The regex brings unneeded complexity. 
https://www.django-rest-framework.org/api-guide/fields/#emailfield

**Comment by oluwagbenga-joloko** [comment]
Please also fix the error message for invalid username ` 'invalid': 'Username should have no spaces or special characters ',` This can be better rephrased to `'Username can only contain letters, numbers, underscores, and hyphens`. Update the regex accordingly too.

File: authors/apps/authentication/serializers.py (hunk 5 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py]: @oluwagbenga-joloko It is that way already or you meant something else.

File: authors/apps/authentication/serializers.py (hunk 6 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py:47]: This is actually better It validates the email

File: authors/apps/authentication/serializers.py (hunk 5 above)
Comment (oluwagbenga-joloko) [reply authors/apps/authentication/serializers.py]: remove the `of` here `password should have at least of 2 uppercase letters`

**Pull request merged**